src/main/java/com/cleanroommc/groovyscript/compat/mods/thaumcraft/arcane/ArcaneRecipeBuilder.java

```diff
@@ -0,0 +1,183 @@
+package com.cleanroommc.groovyscript.compat.mods.thaumcraft.arcane;
+
+import com.cleanroommc.groovyscript.api.GroovyLog;
+import com.cleanroommc.groovyscript.api.IIngredient;
+import com.cleanroommc.groovyscript.compat.mods.thaumcraft.aspect.AspectStack;
+import com.cleanroommc.groovyscript.compat.vanilla.CraftingRecipeBuilder;
+import com.cleanroommc.groovyscript.helper.ingredient.IngredientHelper;
+import com.cleanroommc.groovyscript.registry.ReloadableRegistryManager;
+import it.unimi.dsi.fastutil.chars.Char2ObjectOpenHashMap;
+import net.minecraft.item.crafting.IRecipe;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.fml.common.registry.ForgeRegistries;
+import org.apache.commons.lang3.ArrayUtils;
+import thaumcraft.api.aspects.AspectList;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public abstract class ArcaneRecipeBuilder extends CraftingRecipeBuilder {
+
+    protected String researchKey;
```

Comment:
should default to an empty string meaning research is optional. otherwise, we have to validate it to exist or be an empty string or Thaumic JEI will crash

Reply:
fixed in commit: [a1d68fe](https://github.com/CleanroomMC/GroovyScript/pull/18/commits/a1d68fe07ddf4032ecb44c47c8e0cb9501a1ba00)